### PR TITLE
Add hb_sanitize API and use it on a new tool, hb-sanitize

### PR DESCRIPTION
**Pull request by ebraminio**
Have a look. Maybe it should be called validator, as Apple's ftxvalidator and [the Microsoft one](https://github.com/Microsoft/Font-Validator)?

### CODE REVIEW

File: src/hb-private.hh [hunk 1]
```diff
@@ -123,7 +123,7 @@ extern "C" void  hb_free_impl(void *ptr);
 #endif
 
 #ifndef HB_INTERNAL
-# if !defined(HB_NO_VISIBILITY) && !defined(__MINGW32__) && !defined(__CYGWIN__) && !defined(_MSC_VER) && !defined(__SUNPRO_CC)
+# if !defined(HB_NO_VISIBILITY) && !defined(__MINGW32__) && !defined(__CYGWIN__) && !defined(_MSC_VER) && !defined(__SUNPRO_CC) && !defined(__DJGPP__) && defined(__unix__)
```
Comment (ebraminio) [inline src/hb-private.hh]: As https://stackoverflow.com/q/43287103 we can remove some other if is harmless

**Comment by ebraminio** [comment]
Something fun to check is to see what is the result of sanitization of current fonts so went ahead with `for i in /Library/Fonts/*.tt{c,f} /System/Library/Fonts/*.tt{c,f}; do echo $i; build/util/hb-sanitize "$i" 1>/dev/null; echo $?; done` and saw `/Library/Fonts/Gujarati Sangam MN.ttc` fails on just recent implemented table, `feat`, I should what has went wrong with it.

**Comment by ebraminio** [comment]
About the previous comment, it seems it passes the predefined borders of `feat` table and fonttools also don't like it and not hb-sanitize nor `feat` table implementation fault.

```
ERROR: An exception occurred during the decompilation of the 'feat' table
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/__init__.py", line 409, in __getitem__
    return self.tables[tag]
KeyError: 'feat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/__init__.py", line 424, in __getitem__
    table.decompile(data, self)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otBase.py", line 42, in decompile
    self.table.decompile(reader, font)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otBase.py", line 618, in decompile
    table[conv.name] = conv.read(reader, font, table)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otConverters.py", line 347, in read
    table.decompile(reader, font)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otBase.py", line 614, in decompile
    table[conv.name] = conv.readArray(reader, font, table, countValue)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otConverters.py", line 110, in readArray
    l.append(self.read(reader, font, tableDict))
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otConverters.py", line 347, in read
    table.decompile(reader, font)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otBase.py", line 618, in decompile
    table[conv.name] = conv.read(reader, font, table)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otConverters.py", line 436, in read
    table.decompile(reader, font)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otBase.py", line 614, in decompile
    table[conv.name] = conv.readArray(reader, font, table, countValue)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otConverters.py", line 110, in readArray
    l.append(self.read(reader, font, tableDict))
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otConverters.py", line 347, in read
    table.decompile(reader, font)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otBase.py", line 618, in decompile
    table[conv.name] = conv.read(reader, font, table)
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otConverters.py", line 183, in read
    return reader.readUShort()
  File "/usr/local/lib/python3.6/site-packages/fontTools/ttLib/tables/otBase.py", line 134, in readUShort
    value, = struct.unpack(">H", self.data[pos:newpos])
struct.error: unpack requires a buffer of 2 bytes
```

**Comment by ebraminio** [comment]
PTAL

**Comment by behdad** [comment]
Sanitize is fine as a name.

In general this looks good.  Specifically though, it's not valid to include private headers from the utils. We need to add a `hb_sanitize` API and call it. Maybe give it a try?

**Comment by behdad** [comment]
I suppose it can be `hb_blob_t *hb_sanitize (hb_blob_t *, hb_tag_t)`.

**Comment by ebraminio** [comment]
PTAL

File: TODO [hunk 2]
```diff
@@ -19,8 +19,6 @@ API additions
 
 - Add hb-cairo glue
 
-- Add sanitize API (and a cached version, that saves result on blob user-data)
```
Comment (ebraminio) [inline TODO]: Is the rest something I can do here or should I turn that into a GitHub issue?

File: src/hb-sanitize.cc [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+
+/**
+ * hb_sanitize: Sanitizes a table specified by a tag
+ * @face:
+ * @tag:
+ *
+ *
+ *
+ * Return value: Sanitizer result, nullptr if we don't support that table at all
```
Comment (ebraminio) [inline src/hb-sanitize.cc]: I guess this is not its idiomatic way. Please let me know what should be done instead.

**Comment by ebraminio** [comment]
PTAL

What do you suggest for testing? Maybe some tool to run the tool over the entire the repository and bots font files?

File: src/hb-sanitize.cc [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ * Since: REPLACEME
+ **/
+hb_blob_t* hb_sanitize (hb_face_t* face, hb_tag_t tag)
```
Comment (ebraminio) [inline src/hb-sanitize.cc]: Behdad, your suggested signature had hb_blob_t instead hb_face_t but what should be done with `num_glyphs`? Passing that also thought the API?

File: TODO (hunk 2 above)
Comment (behdad) [reply TODO]: No, don't want the caching.

File: src/hb-sanitize.cc (hunk 3 above)
Comment (behdad) [reply src/hb-sanitize.cc]: Yeah cannot make up my mind so far.

File: src/hb-sanitize.cc [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+ * Return value: Sanitizer result, nullptr if we don't support that table at all
+ * Since: REPLACEME
+ **/
+hb_blob_t *
+hb_sanitize (hb_face_t *face, hb_tag_t tag)
+{
+  unsigned int num_glyphs = hb_face_get_glyph_count (face);
+  hb_blob_t *blob = hb_face_reference_table (face, tag);
+  hb_blob_t *result = nullptr;
+#define DEFINE_SANITIZER(a, b, c, d, table) \
```
Comment (behdad) [inline src/hb-sanitize.cc]: You don't need the tag explicitly. You can use `table::tableTag`.

File: src/hb-sanitize.cc [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  case HB_TAG(a,b,c,d): { \
+    OT::Sanitizer<table> sanitizer; \
+    sanitizer.set_num_glyphs (num_glyphs); \
+    result = sanitizer.sanitize (blob); \
+    break; \
+  }
+  switch (tag)
```
Comment (behdad) [inline src/hb-sanitize.cc]: We should special-case tag 0 to mean sanitise as font file.

File: src/hb-sanitize.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+#ifndef HB_H_IN
+#error "Include <hb.h> instead."
+#endif
+
+#ifndef HB_SANITIZE_H
+#define HB_SANITIZE_H
+
+#include "hb-common.h"
+#include "hb-buffer.h"
+#include "hb-font.h"
```
Comment (behdad) [inline src/hb-sanitize.h]: Why these three? Looks like you need hb-common, hb-face, and hb-blob.

File: util/hb-sanitize.cc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  hb_blob_t *blob = hb_blob_create_from_file (argv[1]);
+  unsigned int faces = 1; // hb_face_count (blob); #1002
+
+  unsigned int fails = 0;
+  for (unsigned int face_index = 0; face_index < faces; ++face_index)
+  {
+    printf ("Face index: %d\n", face_index);
+    hb_face_t *face = hb_face_create (blob, face_index);
+
+    hb_tag_t tags[100];
```
Comment (behdad) [inline util/hb-sanitize.cc]: Use a number like 16 or 32.

File: util/hb-sanitize.cc [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    unsigned int table_count = sizeof (tags) / sizeof (hb_tag_t);
+    hb_face_get_table_tags (face, 0, &table_count, tags);
+
+    for (unsigned int table_index = 0; table_index < table_count; ++table_index)
```
Comment (behdad) [inline util/hb-sanitize.cc]: This is wrong. This doesn't go over all tables.  Look at other places about how to iterate using an api like `get_table_tags` and offsets.

File: util/hb-sanitize.cc [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    {
+      char buf[4];
+      hb_tag_to_string (tags[table_index], buf);
+      hb_blob_t *sanitized = hb_sanitize (face, tags[table_index]);
+
+      if (!sanitized)
+	printf ("%.*s: %s\n", 4, buf, "SKIPPED!");
+      else if (hb_blob_get_length (sanitized))
```
Comment (behdad) [inline util/hb-sanitize.cc]: Check that returned blob is equal to input blob. That's definition of sanitization passing.

File: src/hb-sanitize.cc (hunk 6 above)
Comment (behdad) [reply src/hb-sanitize.cc]: Ok I see that doesn't make sense when you are sending face in.  I'll think about it more...

**Comment by behdad** [comment]
Ok, regarding the API:
- It's unfortunate that we need num_glyphs. That's needed by the AAT::Lookup structure.

Maybe we can add these two APIs:
- `hb_sanitize_file(hb_blob_t, ...)`: This will sanitize as a OpenTypeFontFile.
- `hb_sanitize_table(hb_face_t, tag)`: This does what you have right now.

The results is a returned blob. If the blob is different from the input blob then sanitization had failed. Ouch, ok I see in the second API the caller does not have the blob.

Maybe we should return an enum instead?  Checking for size 0 is not what I would suggest.

**Comment by behdad** [comment]
Also, later, would be nice to have hb-sanitize tool optionally get output filename and save a font with the sanitized tables. That would need the subset_face to become face_builder first.

**Comment by behdad** [comment]
Also, I tested this and it passes even if sanitization failed. That's because sanitizer copies table and fixes things up, if the blob allows that.  Apparently we can do that by making the blob immutable before calling sanitize.

File: src/hb-sanitize.cc (hunk 5 above)
Comment (ebraminio) [reply src/hb-sanitize.cc]: Done

File: src/hb-sanitize.h (hunk 7 above)
Comment (ebraminio) [reply src/hb-sanitize.h]: Done

File: util/hb-sanitize.cc (hunk 8 above)
Comment (ebraminio) [reply util/hb-sanitize.cc]: Done

File: util/hb-sanitize.cc (hunk 9 above)
Comment (ebraminio) [reply util/hb-sanitize.cc]: I think if the input buffer is big enough it go through all of it and fonts table usually don't exceed 32 (or 100?)? I guess I can implement a [double loop] iteration (if that is what you want?) but please give a specific place how that is used. Thanks

**Comment by ebraminio** [comment]
> It's unfortunate that we need num_glyphs. That's needed by the AAT::Lookup structure.

How about passing that also? That feels bad as what if we need other things in future (so we should provide another API maybe), but at least we can provide sanitization API independently from hb_face_t so it can be used on font editor software or something but I am not sure.

**Comment by behdad** [comment]
> It's unfortunate that we need num_glyphs. That's needed by the AAT::Lookup structure.
> 
> How about passing that also? That feels bad as what if we need other things in future (so we should provide another API maybe), but at least we can provide sanitization API independently from hb_face_t so it can be used on font editor software or something but I am not sure.

Yeah I don't like that for the reason you mention.

Alternative is that sanitize would not require this, but we pass down table size to every use of AAT:Lookup and do a runtime check. That was too intrusive to the code, so I did it this way.

**Comment by ebraminio** [comment]
> This is wrong. This doesn't go over all tables. Look at other places about how to iterate using an api like get_table_tags and offsets.

So should I do something like this? https://github.com/freedesktop/libreoffice-core/commit/fa1ec9461e2f78655e9aa17473dba085b0b6015d#diff-edeb470f1693db8a762fae32df27343bR841

**Comment by ebraminio** [comment]
As a part of PR cleanup I am doing these days, as https://github.com/harfbuzz/harfbuzz/pull/950#issuecomment-384802100 and the fact this needs the face itself so can get its glyph count so this patch at least with its current approach feels unlikely to go anywhere, let's close and since it exists on https://github.com/harfbuzz/harfbuzz/blob/master/TODO#L16 we can think about it later. Thanks :)

**Comment by behdad** [comment]
Yeah don't need this. Thanks.

**Pull request closed without merging**